### PR TITLE
explain why unmanaged types aren't marked readonly

### PR DESCRIPTION
**Pull request by BillWagner**
Address this customer question: https://github.com/dotnet/docs/issues/3907#issuecomment-432653335

### CODE REVIEW

File: docs/csharp/write-safe-efficient-code.md [hunk 1]
```diff
@@ -64,6 +64,8 @@ readonly public struct ReadonlyPoint3D
 
 Follow this recommendation whenever your design intent is to create an immutable value type. Any performance improvements are an added benefit. The `readonly struct` clearly expresses your design intent.
 
+The [unamanaged types](language-reference/keywords/sizeof.md) aren't explicitly marked as a `readonly struct` types, but the compiler treats them as such. These types are considered `readonly` from the perspective of the language. We know them to be immutable which is even a stronger guarantee than `readonly` can provide.
```
[the remark this answers is not in the document]
Comment (jaredpar) [reply docs/csharp/write-safe-efficient-code.md]: This paragraph is incorrect. A type which is unmanaged is not treated as `readonly` in any way. Did we suggest that somewhere that I should clarify?

Comment (rpetrusha) [inline docs/csharp/write-safe-efficient-code.md]: some nits:
- as a readonly struct types --> "as readonly struct types", or just "as readonly struct"
- comma between "immutable" and "which"

**Review by rpetrusha — APPROVED** [review]
Two nits about grammar and punctuation, @BillWagner. Once you address the issue of immutable/readonly in .NET Core, you can merge.

File: docs/csharp/write-safe-efficient-code.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/write-safe-efficient-code.md]: @jaredpar I mis-interpreted an email conversation. (I just looked at it again.) I conflated comments about simple types (sbyte, int, long, and such) and readonly ref structs of unmanaged types.

Would the paragraph be correct to state "The simple types `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal` and `bool`" where "unmanaged type" is used above?

Also, can you comment on the comment from @mikedn that `Int32` is now marked as `readonly struct` in .NET Core?

[the remark this answers is not in the document]
Comment (jaredpar) [reply docs/csharp/write-safe-efficient-code.md]: Ah yes. The simple types statement is correct. There are some bugs around this but the intent of the language is to treat them as `readonly`. You can add `enum` to the list as well.

[the remark this answers is not in the document]
Comment (jaredpar) [reply docs/csharp/write-safe-efficient-code.md]: Where is the mikedn comment? Didn't see it on this PR.

[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/write-safe-efficient-code.md]: @jaredpar Here on this issue: https://github.com/dotnet/docs/issues/3907#issuecomment-432658819

Comment (VSadov) [inline docs/csharp/write-safe-efficient-code.md]: The `readonliness` of a struct call is decided here: https://github.com/dotnet/roslyn/blob/60fd993cdab5a4a92490755422984cdfea752be8/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L1687

We special-case 3 methods on `Nullable<T>`, but otherwise we just honor `readonly` annotation on the type that implements the method. Treating primitive types as unconditionally readonly was considered, but I do not think it made it into spec or implementation.

I am not sure it is a very important distinction anyways - If types are truly immutable, cloning is not semantically observable and perf penalty, if any, is very small since these types are very small.
Basically only people that examine actual `IL` might see the difference, which makes this more a curiosity point, not something that would guide you to write your code one way instead of another.

**Comment by BillWagner** [comment]
@jaredpar @VSadov 

I backed out the previous change and added new text that reinforces the recommendation for using `in` arguments with larger ValueTypes that are readonly. The new text emphasizes that the recommendation is only valid for larger value types.

It side-steps the original question, but I think that's the right choice: how the compiler treats immutable simple types, and any change to add the `readonly` modifier to simple types is an implementation detail.

Thoughts?

**Review by jaredpar — APPROVED** [review]
